### PR TITLE
feat: link oidc credentials when login

### PR DESCRIPTION
**Pull request by hperl**
Original PR: https://github.com/ory/kratos/pull/3222

When user tries to login with OIDC for the first time but has already registered before with email/password a credentials identifier conflict may be detected by Kratos. In this case user needs to login with email/password first and then link OIDC credentials on a settings screen.
This PR simplifies UX and allows user to link OIDC credentials to existing account right in the login flow, without
switching to settings flow.

## Related issue(s)
#2727 

## Checklist

- [ x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ x] I have read the [security policy](../security/policy).
- [ x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: selfservice/flow/registration/flow.go [hunk 1]
```diff
@@ -267,3 +262,34 @@ func (f *Flow) GetFlowName() flow.FlowName {
...
+		return nil, err
+	}
+	id, err := uuid.FromString(internalContextOuterFlow.ID)
+	if err != nil {
+		return nil, err
+	}
+	return &id, nil
+}
+
+func (f *Flow) SetOuterLoginFlowID(flowID uuid.UUID) error {
```
Comment (hperl) [inline selfservice/flow/registration/flow.go:288]: I think we should make this a database column instead.

**Comment by hperl** [comment]
@splaunov I cleaned up the linking logic a bit. Thanks again for laying the foundation! PTAL if you like :)

**Review by aeneasr — COMMENTED** [review]
- [ ] Unit / functional tests
- [ ] Normalization working correctly? Test :)
- [ ] Edge case issue with verifiable/recovery addresses
- [ ] Consent step for linking (e.g. checkbox or change of label for submit button)

File: selfservice/flow/login/hook.go [hunk 2]
```diff
@@ -321,3 +328,51 @@ func (e *HookExecutor) PreLoginHook(w http.ResponseWriter, r *http.Request, a *F
...
+// maybeLinkCredentials links the identity with the credentials of the inner context of the login flow.
+func (e *HookExecutor) maybeLinkCredentials(r *http.Request, s *session.Session, i *identity.Identity, f *Flow) error {
+	lc, err := f.DuplicateCredentials()
+	if err != nil {
+		return err
+	} else if lc == nil {
+		return nil
+	}
+
+	if err := e.checkDuplicateCredentialsIdentifierMatch(r.Context(), i.ID, lc.DuplicateIdentifier); err != nil {
```
[the remark this answers is not in the document]
Comment (hperl) [reply selfservice/flow/login/hook.go:341]: Added a test.

File: identity/manager.go [hunk 3]
```diff
@@ -110,12 +110,7 @@ func (m *Manager) Create(ctx context.Context, i *Identity, opts ...ManagerOption
 	return nil
 }
 
-func (m *Manager) findExistingAuthMethod(ctx context.Context, e error, i *Identity) (err error) {
-	if !m.r.Config().SelfServiceFlowRegistrationLoginHints(ctx) {
-		return &ErrDuplicateCredentials{error: e}
-	}
-	// First we try to find the conflict in the identifiers table. This is most likely to have a conflict.
-	var found *Identity
+func (m *Manager) ConflictingIdentity(ctx context.Context, i *Identity) (found *Identity, foundConflictAddress string, err error) {
```
Comment (aeneasr) [inline identity/manager.go:113]: This should come with tests, now that it is being exported to prevent future regressions.

File: selfservice/flow/login/flow.go [hunk 4]
```diff
@@ -231,6 +231,10 @@ func (f *Flow) EnsureInternalContext() {
 	}
 }
 
+func (f *Flow) GetInternalContext() *sqlxx.JSONRawMessage {
+	return &f.InternalContext
```
Comment (aeneasr) [inline selfservice/flow/login/flow.go:235]: Why return a pointer?

File: selfservice/flow/login/hook.go [hunk 5]
```diff
@@ -321,3 +328,51 @@ func (e *HookExecutor) PreLoginHook(w http.ResponseWriter, r *http.Request, a *F
...
+		return err
+	}
+
+	method := strategy.CompletedAuthenticationMethod(ctx)
+	sess.CompletedLoginForMethod(method)
+
+	return nil
+}
+
+func (e *HookExecutor) checkDuplicateCredentialsIdentifierMatch(ctx context.Context, identityID uuid.UUID, match string) error {
```
Comment (aeneasr) [inline selfservice/flow/login/hook.go:365]: Test?

File: selfservice/flow/login/hook.go [hunk 6]
```diff
@@ -321,3 +328,51 @@ func (e *HookExecutor) PreLoginHook(w http.ResponseWriter, r *http.Request, a *F
 
 	return nil
 }
+
+// maybeLinkCredentials links the identity with the credentials of the inner context of the login flow.
+func (e *HookExecutor) maybeLinkCredentials(ctx context.Context, sess *session.Session, ident *identity.Identity, loginFlow *Flow) error {
```
Comment (aeneasr) [inline selfservice/flow/login/hook.go:333]: Test?

File: selfservice/flow/registration/flow.go [hunk 7]
```diff
@@ -208,6 +202,10 @@ func (f *Flow) EnsureInternalContext() {
 	}
 }
 
+func (f *Flow) GetInternalContext() *sqlxx.JSONRawMessage {
+	return &f.InternalContext
```
Comment (aeneasr) [inline selfservice/flow/registration/flow.go:206]: Reconsider pointer here - sqlxx.JSONRawMessage is a byte slice which means that modifying it will change the mem value and not create a copy (I think).

File: schema/errors.go [hunk 8]
```diff
@@ -349,3 +349,13 @@ func NewLoginCodeInvalid() error {
 		Messages: new(text.Messages).Add(text.NewErrorValidationLoginCodeInvalidOrAlreadyUsed()),
 	})
 }
+
+func NewLinkedCredentialsDoNotMatch() error {
+	return errors.WithStack(&ValidationError{
+		ValidationError: &jsonschema.ValidationError{
+			Message:     `linked credentials do not match`,
```
[the remark this answers is not in the document]
Comment (hperl) [reply schema/errors.go:356]: Nothing much, just restart a new flow. Should we do that for the user in this case?

File: selfservice/flow/login/flow.go (hunk 4 above)
Comment (hperl) [reply selfservice/flow/login/flow.go:235]: So that we can mutate the internal context as well.

File: selfservice/flow/registration/flow.go (hunk 7 above)
Comment (hperl) [reply selfservice/flow/registration/flow.go:206]: We do want to change the internal context.

File: schema/errors.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply schema/errors.go:356]: OK - I think you can also show a message "please try again". but creating a new flow would be even nicer!

File: selfservice/flow/login/flow.go (hunk 4 above)
Comment (aeneasr) [reply selfservice/flow/login/flow.go:235]: OK I see - you're using it here:

```
	*flow.GetInternalContext() = bytes
```

I guess that's fine (it works), just a bit unusual (would expect `SetInternalContext()`)

Comment (aeneasr) [reply selfservice/flow/login/flow.go:235]: So what I mean by that is that it's unusual to have pointers to slices. Here's why:

```go
package main

func main() {
	var a []byte
	var b *[]byte
	b = &a
	if b == nil {
		print("nil")
	}
	print((*b)[0]) // panics
}
```

This problem exists here where you check whether the pointer is set, but not whether the internal context is actually set:

```
func SetDuplicateCredentials(flow InternalContexter, creds DuplicateCredentialsData) error {
	if flow.GetInternalContext() == nil {
		flow.EnsureInternalContext()
	}
	bytes, err := sjson.SetBytes(
		*flow.GetInternalContext(),
		internalContextDuplicateCredentialsPath,
		creds,
	)
	if err != nil {
		return err
	}
	*flow.GetInternalContext() = bytes

	return nil
}
```

Since `internalContext` is a []byte per default (and thus nil per default), you'll probably run into some errors here  in `SetBytes`.

File: identity/manager.go (hunk 3 above)
Comment (hperl) [reply identity/manager.go:113]: fixed

File: selfservice/flow/login/flow.go (hunk 4 above)
Comment (hperl) [reply selfservice/flow/login/flow.go:235]: I am calling `flow.EnsureInternalContext()` if the slice is nil to make sure it is always allocated.

In your code example, the panic is not because the pointer points to nothing (it points to `&a`, which is on the stack), but because it is an empty slice and accessing the first element panics.

---

I can still add a SetInternalContext if you want, just to make the code easier to read. WDYT?

File: selfservice/flow/login/hook.go (hunk 5 above)
Comment (hperl) [reply selfservice/flow/login/hook.go:365]: done :)

File: selfservice/flow/login/hook.go (hunk 6 above)
Comment (hperl) [reply selfservice/flow/login/hook.go:333]: done :)

File: selfservice/flow/registration/flow.go (hunk 7 above)
Comment (hperl) [reply selfservice/flow/registration/flow.go]: fixed

File: selfservice/flow/login/flow.go (hunk 4 above)
Comment (hperl) [reply selfservice/flow/login/flow.go]: fixed

File: test/e2e/cypress/integration/profiles/oidc/login/success.spec.ts [hunk 9]
```diff
@@ -1,24 +1,26 @@
...
-      app: "react" as "react",
-      profile: "spa",
-    },
+    // {
+    //   login: react.login,
+    //   registration: react.registration,
+    //   settings: react.settings,
+    //   app: "react" as "react",
+    //   profile: "spa",
+    // },
```
Comment (jonas-jonas) [inline test/e2e/cypress/integration/profiles/oidc/login/success.spec.ts:15]: revert?

**Review by jonas-jonas — CHANGES_REQUESTED** [review]

File: selfservice/flow/duplicate_credentials.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+type DuplicateCredentialsData struct {
+	CredentialsType     identity.CredentialsType
+	CredentialsConfig   sqlxx.JSONRawMessage
+	DuplicateIdentifier string
+}
+
+type InternalContexter interface {
+	EnsureInternalContext()
+	GetInternalContext() sqlxx.JSONRawMessage
```
[the remark this answers is not in the document]
Comment (hperl) [reply selfservice/flow/duplicate_credentials.go:26]: Agreed, but this would conflict with an existing field name. Changing this would change much more code than ignoring the best practices here.

File: test/e2e/cypress/integration/profiles/oidc/login/success.spec.ts (hunk 9 above)
Comment (hperl) [reply test/e2e/cypress/integration/profiles/oidc/login/success.spec.ts:15]: Yes, good catch!

**Review by aeneasr — DISMISSED** [review]

File: selfservice/flow/login/flow.go (hunk 4 above)
Comment (aeneasr) [reply selfservice/flow/login/flow.go]: Yes that's true, what I meant is this:

```
	if flow.GetInternalContext() == nil { // might evaluate to trfalseue because the pointer is not nil but the memory address is nil
		flow.EnsureInternalContext() // would set context to string "{}" but won't because pointer is not-nil even though value is nil
	}
	
	bytes, err := sjson.SetBytes(
		*flow.GetInternalContext(),
		internalContextDuplicateCredentialsPath,
		creds,
	)
	// err is now non-nil because `*flow.GetInternalContext()` is nil which `SetBytes` can't deal with a nil byte slice
```

However, I realized that `SetBytes()` can deal with a nil slice and won't return an error. So this was indeed fine, at the same time I think that the setter is a clearer pattern as it doesn't have this "hidden" nil-pointer versus nil-slice-value meaning.
